### ISSUE TITLE
Server only uses a single path after successfully negotiating multipath

### ISSUE THREAD

**Issue by alexrabi** [private-octopus/picoquic#1707]
After successfully negotiating a multipath connection between the client and server, the client recognizes all paths (path callbacks are triggered) and is able to schedule packets on all available paths. However, while it seems like the server is able to receive packets on all negotiated paths, it never appears to trigger any of the path callbacks (even after using picoquic_enable_path_callbacks), nor are any packets scheduled onto any paths other than the path that was used to set up the initial connection. It is as if the server (or the sending side, at the very least) still considers the connection to be single-path.

Prior to the implementation of multipath-draft-06 (or somewhere thereabout), this was not an issue.

The setup currently looks like this, where the client (c) is connected to the server (s) over two paths (going over r1 and r2), which converge at a gateway (gw), that finally connect to the server:
```
 ____r1____ 
|          |     
c          gw____s
|____r2____|
```

I am running this setup in a Linux environment.

**Comment by huitema** [comment]
The migration to the latest version of the multipath extensions for QUIC was a bit rushed. The server seems to work fine, but we had issues with the client when doing the tests for the IETF. I am currently fixing a first issue (PR #1714), will investigate the path scheduling issue asap.

**Comment by huitema** [comment]
I just did a series of fixes on the multipath code. Load balancing certainly works in my local tests. I would need some more precision on your setup. Are you using 'picoquicdemo', or did you write your own program? Are you able to get a qlog of a short transfert (maybe 1MB)? What are the IP addresses used at the client and at the server?

**Comment by alexrabi** [comment]
The issue was discovered using my own application (a VPN-like multipath tunnel), but the issue seemed to be repeatable with picoquicdemo. After pulling the latest commit (PR #1721), however, I do not seem to be able to run picoquicdemo at all (seems like the server does not actually bind to the correct port, but instead selects a random port?). Anyhow that is a separate issue.

Also, I am not sure if this is related to the current issue, but I am also noticing that the secondary (i.e. non-initial) path is being deleted a few seconds after the connection establishment. This seems related to the MAX_PATH_ID frame after scouring through the qlogs. This also only seems to occur after pulling PR #1721.

The IP addresses are 10.0.3.2 and 10.0.4.2 on the client and 10.0.0.1 on the server.

Anyhow, I modified my application to setup a stream which transfers 1MB from the server to the client at the start of the connection. You can find the qlogs here:

[qlogs.zip](https://github.com/user-attachments/files/16480971/qlogs.zip)

**Comment by huitema** [comment]
When you are testing with picoquicdemo, what command line parameters are you using?

**Comment by alexrabi** [comment]
On the server side:
`picoquicdemo -c certs/cert.pem -k certs/server-key.pem -p 4433 -w ./web`

Client:
`picoquicdemo 10.0.0.1 4433 web/index.html`

And yes, this is without multipath enabled.

**Comment by huitema** [comment]
Regarding your comment "After pulling the latest commit (PR https://github.com/private-octopus/picoquic/pull/1721), however, I do not seem to be able to run picoquicdemo at all". The recent PR #1716 removed the experimental "simple multipath" variant, and that changed the syntax of the "-M" option. Previous values were "-M 1", "-M 2", "-M 3", depending of whether the default multipath option was select (-M 1), or the simple multipath option (-M 2), or either as chosen by the peer (-M 3). But there is no need for a choice anymore, so the syntax changed to just "-M".

**Comment by alexrabi** [comment]
Ah, okay, omitting the "-M 1" seemed to fix that issue. Here are qlogs from picoquicdemo, displaying the same issue (the server seems to only use the secondary path for a single packet in this instance):

[qlogs.zip](https://github.com/user-attachments/files/16487070/qlogs.zip)

The command that was run:
`picoquicdemo -q qlogs -n test -M -A 10.0.3.2/2,10.0.4.2/3 10.0.0.1 4433 /1000000`

**Comment by huitema** [comment]
Thanks for the logs. There is an anomaly: the client sends path challenges on the new paths, receives a challenge and a responses from the server, but does not respond to the server's challenge. Since there is no response, the server does not consider that the path is established, so does not send any data on it. I will investigate why.

**Comment by huitema** [comment]
The client's  log shows this:
~~~
[35288, 1, "transport", "packet_received", { "packet_type": "1RTT", "header": { "packet_size": 1252, "packet_number": 0, "dcid": "a0b55d1549343ae5" }, "frames": [{ 
    "frame_type": "path_challenge", "path_challenge": "77df6a4012c14538"}, { 
    "frame_type": "path_response", "path_response": "afcf7269a9a6dfcd"}, { 
    "frame_type": "padding"}]}],
[35288, 1, "recovery", "metrics_updated", {"cwnd": 15360,"pacing_rate": 10016000000,"bytes_in_flight": 1252,"smoothed_rtt": 9484,"min_rtt": 9484,"latest_rtt": 9484,"app_limited": 0}],
[35304, 1, "transport", "packet_sent", { "packet_type": "1RTT", "header": { "packet_size": 39, "packet_number": 1, "dcid": "0dc066ef662910ad", "quic_bit": 0 }, "frames": [{ 
    "frame_type": "time_stamp", "time_stamp": 4413}, { 
    "frame_type": "mp_ack", "path_id": 0, "ack_delay": 0, "acked_ranges": [[12, 38]], "ect0": 0, "ect1": 39, "ce": 0}, { 
    "frame_type": "mp_ack", "path_id": 1, "ack_delay": 0, "acked_ranges": [[0, 0]], "ect0": 0, "ect1": 1, "ce": 0}, { 
    "frame_type": "padding"}]}],
~~~
The client receives the server's packet, containing the server's challenge and a response to the client's challenge. The client almost immediately sends a response containing acknowledgements, including an acknowledgement of the server's packet that contained the challenge. But it does not send the challenge response. This implies that the challenge was somehow ignored. If I look at the code, this only happens if the IP source and destination addresses do not match the addresses at which the challenge was sent. I am not 100% sure, but this test was probably added with multipath-draft-06.

I have verified that the multipath code works as expected in unit tests, but the unit tests use a network simulation instead of the socket code. I have also verified that the multipath code works as expected in the "simulation" work, in which the two client paths differ by their port numbers. It is possible that there is something wrong in the case when the two client paths differ by their IP address.

I am adding diagnostic code in PR #1724. Can you ran that code in your environment and produce another set of qlogs?

**Comment by alexrabi** [comment]
Sure thing: [qlogs.zip](https://github.com/user-attachments/files/16491644/qlogs.zip)

**Comment by huitema** [comment]
I think I understand what is happening! The log has this line:
~~~
[37963, 1, "transport", "packet_received", { "packet_type": "1RTT", "header": { "packet_size": 1252, "packet_number": 0, "dcid": "5d9c129fb0966777" }, "frames": [{ 
    "frame_type": "path_challenge", "path_challenge": "6939656cd47a9f34"}, { 
    "frame_type": "path_response", "path_response": "23c21a25bfe753fa"}, { 
    "frame_type": "padding"}]}],
[37992, 0, "info", "message", { "message": "Path challenge[1] from 10.0.0.1:4433 to 10.0.3.2:43768 ignored, wrong addresses, expected 10.0.0.1:4433 - 10.0.3.2:0.?"}],
~~~
That's a bug in the implementation of the "-A" option in picoquic demo. The packets are being sent on the same UDP sockets as path 0, using the same port number (43768), instead of the unspecified value 0. It is very late, so no immediate fix now, but I should be able to have a fix tomorrow.

**Comment by alexrabi** [comment]
Indeed, it does appear that all packets from the client use the same port number for any path, using both picoquicdemo and using my own application.

**Comment by huitema** [comment]
I have a PR in progress (#1725) to fix the port number issue. Could you try the code in branch debug-multipath-draft-10 with your setup and verify that the issue is fixed?

**Comment by huitema** [comment]
Closed by mistake. Waiting for confirmation that this is really fixed.

**Comment by alexrabi** [comment]
PR #1725 does appear to fix the issue (for picoquicdemo, that is). Now my concern would be to replicate this fix for my own application; the fix in PR #1725 uses data that is internal to picoquic (i.e. picoquic_internal.h) when accessing the port number of the initial path. In other words, the actual API when using picoquic as an external library does not allow for the same kind of fix. 

Perhaps an update to the API is in order, to make it possible to ask picoquic for information on the address/port of the initial path (or any path, really), which I believe is not currently possible? Another idea would be to add an alternative to picoquic_probe_new_path_ex(), which fills in the needed information automatically, since I imagine this would be a pretty common use-case for MP-QUIC?

**Comment by huitema** [comment]
PR #1726 adds an API to retrieve the addresses used by a path. Will that work?

**Comment by alexrabi** [comment]
It is late here, but I will check out the changes and see if I run into any issues first thing tomorrow.

**Comment by alexrabi** [comment]
Alright, so after doing some testing with the new API, it does work, _sort of_. However, if I bind the client to an address/port using `picoquic_set_local_addr` the path challenge still fails, since `picoquic_get_path_addr` returns the bound address while the challenge expects another:

```
[69467, 0, "transport", "datagram_received", { "byte_length": 1252, "addr_to" : {"ip_v4": "10.0.4.2", "port_v4":1234}}],
[69467, 1, "transport", "packet_received", { "packet_type": "1RTT", "header": { "packet_size": 1252, "packet_number": 0, "dcid": "d07fc8e86a7b9680" }, "frames": [{ 
    "frame_type": "path_challenge", "path_challenge": "e3efd691d20d9ff7"}, { 
    "frame_type": "path_response", "path_response": "e2ece52dfd858558"}, { 
    "frame_type": "padding"}]}],
[69538, 0, "info", "message", { "message": "Path challenge[1] from 10.0.0.1:5000 to 10.0.4.2:1234 ignored, wrong addresses, expected 10.0.0.1:5000 - 10.0.4.2:53764.?"}],
```

**Comment by huitema** [comment]
Damn. I probably need to take a different tack, because the test is not robust. Something like, "accept the challenge if the local address IP does not match but the IP address matches and the port number was 0", and then set the port number to the received value.

**Comment by huitema** [comment]
In the "bound" test, how are you managing the sockets? Is your socket bound to (10.0.3.2:1234) or to (0.0.0.0:1234)? Or do you have several sockets?

**Comment by huitema** [comment]
I just pushed a new commit in PR #1726, to loosen the challenge path matching test and remove the forcing of port number in picoquicdemo. Can you remove the code that sets the port for the outgoing path and try again?

**Comment by alexrabi** [comment]
The sockets are handled by the built-in `picoquic_packet_loop`, which creates one socket that will be bound to (0.0.0.0:1234), and another socket with a random port. Relaxing the test indeed works, now both paths appear to be established correctly.

**Comment by huitema** [comment]
OK, so I will merge that, because it is the simplest solution. I wonder whether I should maybe just get rid of the challenge accept text altogether. It is meant to solve scenarios in which both paths try to establish "path number X" at the same time. In those scenarios, we want the client to win. So what we aim for is:

* client starts challenge with Source=A, Dest=B
* server starts a parallel challenge with Source=B, Dest=C
* server receives client challenge and accepts it, which results in changing path addresses to Source=B, Dest=A
* client does not change the source and dest address of the path.

Then there is a small security issue of accepting packets with random addresses, because they can be sent by third parties. But the test is not a good way to solve that.

I think there is also an issue with the socket loop. If the port number is set to 0, it should use the default socket, the one bound to 1234 in your example. But it is good that we see the issue, it shows that the test in the path challenge was too strict.

**Comment by huitema** [comment]
Looking back at this code. The address checks date from 2018, i.e., before we used multipath. At the time, the checks were there because the "path" was identified by the address pair. That changed 3 months ago (draft 6) as the path is now identified by its unique path ID -- unless multipath is not negotiated. I think that I have enough information to fix the code.

**Comment by alexrabi** [comment]
Sounds like a good plan to me.

I agree that the socket loop probably needs fixing: had it used the socket bound to 0.0.0.0:1234 as default then the path challenge would not have been an issue for the bound test. However, as you say, the test may still be too strict with the current checks.

**Comment by huitema** [comment]
Fixed in PR #1726
